src/wallet/gtest/test_wallet_zkeys.cpp

```diff
@@ -331,6 +331,163 @@ TEST(WalletZkeysTest, WriteViewingKeyDirectToDB) {
+ * - Were serialized with an incorrect checksum (check for failure)
+ */
+TEST(WalletZkeysTest, ZKeyChecksumTests) {
+    SelectParams(CBaseChainParams::TESTNET);
+
+    // Get temporary and unique path for file.
+    // Note: / operator to append paths
+    fs::path pathTemp = fs::temp_directory_path() / fs::unique_path();
+    fs::create_directories(pathTemp);
+    mapArgs["-datadir"] = pathTemp.string();
+
+    bool fFirstRun;
+    CWallet wallet("wallet-checksums.dat");
+    LOCK(wallet.cs_wallet);
+    ASSERT_EQ(DB_LOAD_OK, wallet.LoadWallet(fFirstRun));
+
+    // No default CPubKey set
+    ASSERT_TRUE(fFirstRun);
+
+    // create random key and add it to database directly, bypassing wallet
+    auto sk = libzcash::SproutSpendingKey::random();
+    auto addr = sk.address();
+    int64_t now = GetTime();
+    CKeyMetadata meta(now);
+    WalletTestDBWrapper db("wallet-checksums.dat");
+    ASSERT_TRUE(db.WriteZKeyNoChecksum(addr, sk, meta));
+
+    // wallet should not be aware of key
+    ASSERT_FALSE(wallet.HaveSproutSpendingKey(addr));
+
+    // load the wallet again
+    ASSERT_EQ(DB_LOAD_OK, wallet.LoadWallet(fFirstRun));
+
+    // wallet can now see the spending key
+    ASSERT_TRUE(wallet.HaveSproutSpendingKey(addr));
+
+    // Add a zkey with a bad checksum
+    auto sk2 = libzcash::SproutSpendingKey::random();
+    auto addr2 = sk2.address();
+    ASSERT_TRUE(db.WriteZKeyWrongChecksum(addr2, sk2, meta));
+
+    // load the wallet again -- should fail with bad checksum
+    ASSERT_NE(DB_LOAD_OK, wallet.LoadWallet(fFirstRun));
+}
+
+/**
+ * This test covers reading a SaplingZKey from the WalletDB that:
+ * - Were not serialized with a checksum (check for succuss)
+ * - Were serialized with an incorrect checksum (check for failure)
+ */
+TEST(WalletZkeysTest, ZKeySaplingChecksumTests) {
+    SelectParams(CBaseChainParams::TESTNET);
+
+    // Get temporary and unique path for file.
+    // Note: / operator to append paths
+    fs::path pathTemp = fs::temp_directory_path() / fs::unique_path();
+    fs::create_directories(pathTemp);
+    mapArgs["-datadir"] = pathTemp.string();
+
+    CWallet walletSapling("wallet-checksums-sapling.dat");
+    CWallet walletDummy("wallet-dummy.dat");
+    bool fFirstRun;
+    ASSERT_EQ(DB_LOAD_OK, walletSapling.LoadWallet(fFirstRun));
+    ASSERT_EQ(DB_LOAD_OK, walletDummy.LoadWallet(fFirstRun));
+
+    // Generate random key from dummy wallet. GenerateNewSpaplingZKey() automatically databases results, so the key
+    // has to be generated from a different wallet.
+    walletDummy.GenerateNewSeed();
+    auto address = walletDummy.GenerateNewSaplingZKey();
+
+    // Generate a diversified address different to the default
+    // If we can't get an early diversified address, we are very unlucky
+    libzcash::SaplingExtendedSpendingKey extsk;
+    EXPECT_TRUE(walletDummy.GetSaplingExtendedSpendingKey(address, extsk));
+    blob88 diversifier;
+    diversifier.begin()[0] = 10;
```

Comment:
`extsk` is random I think, so why is this diversifier guaranteed to be valid?

Reply:
`blob88 diversifier; diversifier.begin()[0] = 10;` appears to be an unused code fragment that I did not remove in my final cleanup (appears below on line 484 as well). I will push a commit to remove those. 

Is there any reason to be concerned that using existing wallet functionality to create the new Sapling key and extract the extended spending key would produce inconsistent results when databasing and causing these tests to fail or pass when they should not? From my understanding of the code, it seems like so long as the values do not already exist in the walletdb they will be stored and retrieved from the walletdb as expected.